etl/steps/data/garden/energy/2022-12-13/uk_historical_electricity.py

```diff
@@ -0,0 +1,213 @@
+    df_beis["solar_generation"] = 0
+    df_beis["wind_generation"] = 0
+    # Similarly, given that in the BEIS dataset there is no data about other renewable sources (apart from hydro, solar
+    # and wind), we can assume that the contribution from other renewables is zero.
+    df_beis["other_renewables_generation"] = 0
+    # And ensure these new columns do not have any values after the electricity mix data begins.
+    df_beis.loc[
+        df_beis["year"] >= solar_or_wind_first_year,
+        ["solar_generation", "wind_generation", "other_renewables_generation"],
+    ] = np.nan
+
+    # BEIS data on fuel input gives raw energy, but we want electricity generation (which is less, given the
+    # inefficiencies of the process of burning fossil fuels).
+    # They also include a variable on "implied efficiency", which they obtain by dividing the input energy by the total
+    # electricity generation.
+    # We multiply the raw energy by the efficiency to have an estimate of the electricity generated by each fossil fuel.
+    # This only affects data prior to the beginning of the electricity mix's data (which is 1965 for renewables and
+    # nuclear, and 1985 for the rest).
+    for source in ["coal", "oil", "gas"]:
+        df_beis[f"{source}_generation"] *= df_beis["implied_efficiency"]
+
+    # Drop other unnecessary columns.
+    df_beis = df_beis.drop(columns="implied_efficiency")
+
+    # Combine BEIS and electricity mix data.
+    df_combined = dataframes.combine_two_overlapping_dataframes(
+        df1=df_elec, df2=df_beis, index_columns=["country", "year"]
+    )
+
+    # Add an index and sort conveniently.
+    df_combined = df_combined.set_index(["country", "year"]).sort_index().sort_index(axis=1)
+
+    return cast(pd.DataFrame, df_combined)
+
+
+def run(dest_dir: str) -> None:
+    #
+    # Load data.
+    #
+    # Read all required datasets.
+    ds_beis = catalog.Dataset(UK_BEIS_DATASET_PATH)
+    ds_elec = catalog.Dataset(ELECTRICITY_MIX_DATASET_PATH)
+
+    # Gather all required tables from all datasets.
+    tb_beis = ds_beis[UK_BEIS_TABLE_NAME]
+    tb_elec = ds_elec[ELECTRICITY_MIX_TABLE_NAME]
+
+    # Create convenient dataframes.
+    df_beis = pd.DataFrame(tb_beis).reset_index()
+    df_elec = pd.DataFrame(tb_elec).reset_index()
+    #
+    # Process data.
+    #
+    # Prepare electricity mix data.
+    df_elec = prepare_electricity_mix_data(df_elec=df_elec)
+    # Prepare BEIS data.
+    df_beis = prepare_beis_data(df_beis=df_beis)
+
+    # Combine BEIS and electricity mix data.
+    df_combined = combine_beis_and_electricity_mix_data(df_beis=df_beis, df_elec=df_elec)
+
+    # Create a new table with combined data (and no metadata).
+    tb_combined = catalog.Table(df_combined)
+
+    #
+    # Save outputs.
+    #
+    ds_garden = catalog.Dataset.create_empty(dest_dir)
+    # Get the rest of the metadata from the yaml file.
+    ds_garden.metadata.update_from_yaml(METADATA_PATH, if_source_exists="replace")
```

Comment:
You could use the new method: `ds_garden.update_metadata(METADATA_PATH)`. This also affects the building of table `tb_combined`.

Note that you should call this function after adding the tables to the dataset.